File: sdk/iris_otel/instrumentation.py
from __future__ import annotations

import asyncio
import functools
import os
import time
from contextlib import contextmanager

from opentelemetry.trace import Status, StatusCode

from .metrics import get_duration_histogram, get_token_histogram
from .tracer import get_tracer

# The GenAI semantic conventions deliberately keep prompt/response content
# out of default telemetry for privacy; capture is opt-in only.
CAPTURE_CONTENT = os.environ.get("IRIS_CAPTURE_CONTENT", "false").lower() == "true"


@contextmanager
def observe(name: str = "invoke_agent", **attributes):
    """Top-level span for one agent operation, per the `invoke_agent` OTel GenAI span kind.

    A plain sync context manager works fine even inside `async def` code --
    span start/end is local bookkeeping, not I/O, so there's no async variant.
    """
    tracer = get_tracer()
    with tracer.start_as_current_span(name) as span:
        for key, value in attributes.items():
            span.set_attribute(key, value)
        yield span


def trace_llm_call(
    model: str | None = None,
    extract_usage=None,
    extract_finish_reasons=None,
    extract_messages=None,
    system_instructions: str | None = None,
):
    """Wrap an LLM call in a `chat` span with `gen_ai.*` attributes (OTel GenAI conventions),
    and record the real OTel Metrics the conventions also specify --
    gen_ai.client.operation.duration and gen_ai.client.token.usage histograms
    -- not just span attributes a dashboard has to aggregate after the fact.

    Works on both sync and async (`async def`) target functions -- the
    wrapped call is awaited if the original was a coroutine function, since
    an agent doing multi-round tool use typically calls the client async.

    `extract_usage(result) -> (input_tokens, output_tokens)`,
    `extract_finish_reasons(result) -> str | list[str]`, and
    `extract_messages(result) -> Any` let callers adapt to their provider's
    response shape; see `presets.py` for ready-made Anthropic extractors.
    """

    def decorator(func):
        def _set_request_attrs(span, kwargs):
            resolved_model = model or kwargs.get("model")
            if resolved_model:
                span.set_attribute("gen_ai.request.model", resolved_model)

            if CAPTURE_CONTENT:
                messages = kwargs.get("messages")
                if messages is not None:
                    span.set_attribute("gen_ai.input.messages", str(messages))
                # system_instructions covers a static, decoration-time prompt;
                # kwargs["system"] covers one built dynamically per call (e.g.
                # a prompt that interpolates per-request state) -- prefer the
                # explicit static one if both happen to be given.
                resolved_system = system_instructions or kwargs.get("system")
                if resolved_system:
                    span.set_attribute("gen_ai.system_instructions", str(resolved_system))

            return resolved_model

        def _set_response_attrs(span, result, resolved_model, duration_s):
            input_tokens = output_tokens = None
            if extract_usage:
                input_tokens, output_tokens = extract_usage(result)
                if input_tokens is not None:
                    span.set_attribute("gen_ai.usage.input_tokens", input_tokens)
                if output_tokens is not None:
                    span.set_attribute("gen_ai.usage.output_tokens", output_tokens)

            if extract_finish_reasons:
                span.set_attribute("gen_ai.response.finish_reasons", extract_finish_reasons(result))

            if CAPTURE_CONTENT and extract_messages:
                span.set_attribute("gen_ai.output.messages", str(extract_messages(result)))

            metric_attrs = {"gen_ai.request.model": resolved_model} if resolved_model else {}
            get_duration_histogram().record(duration_s, metric_attrs)
            if input_tokens is not None:
                get_token_histogram().record(input_tokens, {**metric_attrs, "gen_ai.token.type": "input"})
            if output_tokens is not None:
                get_token_histogram().record(output_tokens, {**metric_attrs, "gen_ai.token.type": "output"})

        if asyncio.iscoroutinefunction(func):

            @functools.wraps(func)
            async def async_wrapper(*args, **kwargs):
                tracer = get_tracer()
                with tracer.start_as_current_span("chat") as span:
                    resolved_model = _set_request_attrs(span, kwargs)
                    start = time.perf_counter()
                    try:
                        result = await func(*args, **kwargs)
                    except Exception as exc:
                        span.record_exception(exc)
                        span.set_status(Status(StatusCode.ERROR, str(exc)))
                        raise
                    duration_s = time.perf_counter() - start
                    _set_response_attrs(span, result, resolved_model, duration_s)
                    return result

            return async_wrapper

        @functools.wraps(func)
        def sync_wrapper(*args, **kwargs):
            tracer = get_tracer()
            with tracer.start_as_current_span("chat") as span:
                resolved_model = _set_request_attrs(span, kwargs)
                start = time.perf_counter()
                try:
                    result = func(*args, **kwargs)
                except Exception as exc:
                    span.record_exception(exc)
                    span.set_status(Status(StatusCode.ERROR, str(exc)))
                    raise
                duration_s = time.perf_counter() - start
                _set_response_attrs(span, result, resolved_model, duration_s)
                return result

        return sync_wrapper

    return decorator

File: sdk/js/src/instrumentation.ts
import { performance } from "node:perf_hooks";
import { SpanStatusCode, type Attributes, type Span } from "@opentelemetry/api";
import { getDurationHistogram, getTokenHistogram } from "./metrics.js";
import { getTracer } from "./tracer.js";

// The GenAI semantic conventions deliberately keep prompt/response content
// out of default telemetry for privacy; capture is opt-in only.
const CAPTURE_CONTENT = (process.env.IRIS_CAPTURE_CONTENT ?? "false").toLowerCase() === "true";

/**
 * Top-level span for one agent operation, per the `invoke_agent` OTel GenAI span kind.
 *
 * `fn` runs with the new span active (so a `traceLlmCall`-wrapped call inside
 * it becomes a child span automatically), and its return value/throw
 * propagates through unchanged. Works for sync or async `fn`.
 */
export async function observe<T>(
  name: string,
  attributes: Attributes,
  fn: (span: Span) => T | Promise<T>,
): Promise<T> {
  const tracer = getTracer();
  return tracer.startActiveSpan(name, async (span) => {
    for (const [key, value] of Object.entries(attributes)) {
      span.setAttribute(key, value as never);
    }
    try {
      return await fn(span);
    } catch (err) {
      span.recordException(err as Error);
      span.setStatus({ code: SpanStatusCode.ERROR, message: (err as Error)?.message });
      throw err;
    } finally {
      span.end();
    }
  });
}

type ExtractUsage<TResult> = (result: TResult) => [number | undefined, number | undefined] | undefined;
type ExtractFinishReasons<TResult> = (result: TResult) => string | string[] | undefined;
type ExtractMessages<TResult> = (result: TResult) => unknown;

export interface TraceLlmCallOptions<TArgs, TResult> {
  model?: string;
  extractUsage?: ExtractUsage<TResult>;
  extractFinishReasons?: ExtractFinishReasons<TResult>;
  extractMessages?: ExtractMessages<TResult>;
  systemInstructions?: string;
  /** Pull `model`/`messages`/`system` for span attributes out of the call args (default: first arg, as an options-object-style SDK call like Anthropic's `messages.create({...})`). */
  getRequestFields?: (args: TArgs) => { model?: string; messages?: unknown; system?: unknown };
}

const defaultGetRequestFields = (args: unknown[]): { model?: string; messages?: unknown; system?: unknown } => {
  const first = args[0];
  if (first && typeof first === "object") {
    const obj = first as Record<string, unknown>;
    return { model: obj.model as string | undefined, messages: obj.messages, system: obj.system };
  }
  return {};
};

/**
 * Wrap an LLM call in a `chat` span with `gen_ai.*` attributes (OTel GenAI conventions),
 * and record the real OTel Metrics the conventions also specify --
 * gen_ai.client.operation.duration and gen_ai.client.token.usage histograms
 * -- not just span attributes a dashboard has to aggregate after the fact.
 *
 * `extractUsage(result)`, `extractFinishReasons(result)`, and
 * `extractMessages(result)` let callers adapt to their provider's response
 * shape; see `presets/anthropic.ts` for ready-made Anthropic extractors.
 */
export function traceLlmCall<TArgs extends unknown[], TResult>(
  options: TraceLlmCallOptions<TArgs, TResult>,
  fn: (...args: TArgs) => Promise<TResult>,
): (...args: TArgs) => Promise<TResult> {
  const getRequestFields = options.getRequestFields ?? (defaultGetRequestFields as (args: TArgs) => ReturnType<typeof defaultGetRequestFields>);

  return async (...args: TArgs): Promise<TResult> => {
    const tracer = getTracer();
    return tracer.startActiveSpan("chat", async (span) => {
      const requestFields = getRequestFields(args);
      const resolvedModel = options.model ?? requestFields.model;
      if (resolvedModel) {
        span.setAttribute("gen_ai.request.model", resolvedModel);
      }

      if (CAPTURE_CONTENT) {
        if (requestFields.messages !== undefined) {
          span.setAttribute("gen_ai.input.messages", JSON.stringify(requestFields.messages));
        }
        // systemInstructions covers a static, wrap-time prompt; the call's
        // own `system` field covers one built dynamically per call -- prefer
        // the explicit static one if both happen to be given.
        const resolvedSystem = options.systemInstructions ?? requestFields.system;
        if (resolvedSystem !== undefined) {
          span.setAttribute("gen_ai.system_instructions", String(resolvedSystem));
        }
      }

      const start = performance.now();
      try {
        const result = await fn(...args);
        const durationS = (performance.now() - start) / 1000;

        let inputTokens: number | undefined;
        let outputTokens: number | undefined;
        if (options.extractUsage) {
          [inputTokens, outputTokens] = options.extractUsage(result) ?? [undefined, undefined];
          if (inputTokens !== undefined) span.setAttribute("gen_ai.usage.input_tokens", inputTokens);
          if (outputTokens !== undefined) span.setAttribute("gen_ai.usage.output_tokens", outputTokens);
        }

        if (options.extractFinishReasons) {
          const finishReasons = options.extractFinishReasons(result);
          if (finishReasons !== undefined) {
            span.setAttribute("gen_ai.response.finish_reasons", finishReasons);
          }
        }

        if (CAPTURE_CONTENT && options.extractMessages) {
          span.setAttribute("gen_ai.output.messages", JSON.stringify(options.extractMessages(result)));
        }

        const metricAttrs: Attributes = resolvedModel ? { "gen_ai.request.model": resolvedModel } : {};
        getDurationHistogram().record(durationS, metricAttrs);
        if (inputTokens !== undefined) {
          getTokenHistogram().record(inputTokens, { ...metricAttrs, "gen_ai.token.type": "input" });
        }
        if (outputTokens !== undefined) {
          getTokenHistogram().record(outputTokens, { ...metricAttrs, "gen_ai.token.type": "output" });
        }

        return result;
      } catch (err) {
        span.recordException(err as Error);
        span.setStatus({ code: SpanStatusCode.ERROR, message: (err as Error)?.message });
        throw err;
      } finally {
        span.end();
      }
    });
  };
}
